### PR TITLE
feat(hash): add poseidon hash

### PR DESCRIPTION
**Pull request by LucasLvy**
# Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- Bugfix
- Feature
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no API changes)
- Build-related changes
- Documentation content changes
- Testing
- Other (please describe):

## What is the current behavior?

Resolves: #NA

## What is the new behavior?

-
-
-

### CODE REVIEW

File: crates/starknet-types-core/src/hash/poseidon.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+        assert_eq!(
+            Poseidon::hash(&x, &y),
+            Felt::from_hex("0x67c6a2e2d0c7867f97444ae17956dbc89d40ad22255bb06f5f6c515958926ed")
+                .unwrap()
+        );
+    }
+
+    #[test]
+    fn test_pedersen_hash_array() {
```
Comment (tdelabro) [inline crates/starknet-types-core/src/hash/poseidon.rs:49]: poseidon not pedersen

File: crates/starknet-types-core/src/hash/poseidon.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    fn hash(felt_0: &Felt, felt_1: &Felt) -> Felt {
+        Felt(PoseidonCairoStark252::hash(&felt_0.0, &felt_1.0))
+    }
+    /// Computes the Poseidon hash of an array of Felts, as defined
+    /// in <https://docs.starknet.io/documentation/architecture_and_concepts/Cryptography/hash-functions/#poseidon_array_hash.>
+    fn hash_array(felts: &[Felt]) -> Felt {
+        Felt(PoseidonCairoStark252::hash_many(unsafe {
+            core::mem::transmute::<&[Felt], &[FieldElement<Stark252PrimeField>]>(felts)
+        }))
+    }
```
Comment (tdelabro) [inline crates/starknet-types-core/src/hash/poseidon.rs:25]: Don't use transmute, too much weird behaviours (alignment and so on).
You can take inspiration from this snippet of madara:
```rust
fn convert_calldata(data: &[Felt252Wrapper]) -> &[FieldElement] {
    // Non-copy but less dangerous than transmute
    // https://doc.rust-lang.org/std/mem/fn.transmute.html#alternatives
    unsafe { core::slice::from_raw_parts(data.as_ptr() as *const FieldElement, data.len()) }
}
```

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/starknet-types-core/src/hash/poseidon.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        Felt(PoseidonCairoStark252::hash(&felt_0.0, &felt_1.0))
+    }
+    /// Computes the Poseidon hash of an array of Felts, as defined
+    /// in <https://docs.starknet.io/documentation/architecture_and_concepts/Cryptography/hash-functions/#poseidon_array_hash.>
+    fn hash_array(felts: &[Felt]) -> Felt {
+        // Non-copy but less dangerous than transmute
+        // https://doc.rust-lang.org/std/mem/fn.transmute.html#alternatives
+        Felt(PoseidonCairoStark252::hash_many(unsafe {
+            core::slice::from_raw_parts(
+                felts.as_ptr() as *const FieldElement<Stark252PrimeField>,
```
Comment (tdelabro) [inline crates/starknet-types-core/src/hash/poseidon.rs:26]: I think for it to be entierly safe you need `#[repr(transparent)]` ontop of `Felt`

Comment (LucasLvy) [reply crates/starknet-types-core/src/hash/poseidon.rs:26]: it is already

Comment (tdelabro) [reply crates/starknet-types-core/src/hash/poseidon.rs:26]: oh sorry I missed it

**Review by tdelabro — APPROVED** [review]

**Pull request merged**